### PR TITLE
[msan] Unpoison indirect outputs for userspace when -msan-handle-asm-conservative is specified

### PR DESCRIPTION
**Pull request by MaskRay**
KMSAN defaults to `msan-handle-asm-conservative`, which inserts
`__msan_instrument_asm_store` calls to unpoison indirect outputs in
inline assembly (e.g. `=m` constraints in source).

```c
unsigned f() {
  unsigned v;
  // __msan_instrument_asm_store unpoisons v before invoking the asm.
  asm("movl $1,%0" : "=m"(v));
  return v;
}
```

Extend the mechanism to userspace, but require explicit
`-mllvm -msan-handle-asm-conservative` for experiments for now.

As
https://docs.kernel.org/dev-tools/kmsan.html#inline-assembly-instrumentation
says, this approach may mask certain errors (an indirect output may not
actually be initialized), but it also helps to avoid a lot of false
positives.

Link: https://github.com/google/sanitizers/issues/192

### CODE REVIEW

**Comment by MaskRay** [comment]
Ping:)

File: llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp [hunk 1]
```diff
@@ -4557,7 +4561,13 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
       return;
     Value *SizeVal =
       IRB.CreateTypeSize(MS.IntptrTy, DL.getTypeStoreSize(ElemTy));
-    IRB.CreateCall(MS.MsanInstrumentAsmStoreFn, {Operand, SizeVal});
+    if (MS.CompileKernel) {
+      IRB.CreateCall(MS.MsanInstrumentAsmStoreFn, {Operand, SizeVal});
+    } else {
+      auto [ShadowPtr, _] =
+          getShadowOriginPtrUserspace(Operand, IRB, IRB.getInt8Ty(), Align(1));
+      IRB.CreateAlignedStore(getCleanShadow(ElemTy), ShadowPtr, Align(1));
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:4569]: Any reason we use the the Aligned variant? Align(1) should be the same as unspecified alignment, so we can just use `CreateStore`, right?

Comment (MaskRay) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:4569]: ElemTy` utilizes `elementtype` (`ptr elementtype(i32) @id1` => i32, which does not encode the alignment).
An unspecified alignment uses the default, `ElemTy`'s alignment, which is 4 in the i32 case.
However, if the element type is actually unaligned, `CreateStore` could cause a shadow misalignment issue on certain architectures using `-mstrict-align`.

The tests specifically write `, align 1`

Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:4569]: Can you add this as a comment? This is quite subtle and very likely lost on cursory readers.

Comment (MaskRay) [reply llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:4571]: Done:)

**Review by vitalybuka — APPROVED** [review]

**Pull request merged**